### PR TITLE
Adds effect_on_condition, allowing json scriptable events to happen at any time.

### LINKED ISSUE #28290: Implement general purpose timer queues
## Timer Queues
CDDA needs a standard API to run a task/function at some specific future date.  This API would also need the ability to reschedule a task and cancel a task.

The immediate obvious use would be off-map NPC behavior, but I'm sure there's lots of other uses.

### Details TBD
I'm fuzzy on the implementation details.  Someone with a better idea should file this in.

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
Infrastructure "Add effect_on_conditions allowing use of dialog conditions/effects outside of dialog"

#### Purpose of change
Fix #28290
First part of #42786
#42786 is way too big for merge so this is the barebones version of it. 
Adds effect_on_conditions to game

#### Describe the solution
This is the skeleton of the effect_on_condition work.  

Added a new generic factory effect_on_condition.
This allows running pairs of dialog conditions and effects as often as desired via json.  The included documentation explains the exact syntax.

This PR contains examples in the documentation but no usages yet.  Future PR's will break up those parts of #42786 including the weather system, item activation, monster attacks/death and more.

### CODE REVIEW

File: doc/EFFECT_ON_CONDITION.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+### effect_on_condition
+An effect_on_condition is an object allowing the combination of dialog conditions and effects and their usage outside of a dialog.  When invoked they will test their condition and on a pass then cause their effect. They can be activated automatically with any given frequency.  effect_on_conditions use the npc dialog conditions and effects syntax and those allow checking or effecting an npc, which an effect_on_condition does not have.  Doing so is nonsensical, unsupported, and should not be used in an effect_on_condition.
```
[the remark this answers is not in the document]
Comment (Ramza13) [reply doc/EFFECT_ON_CONDITION.md]: Well I just tried, let me know if it's still unclear.

File: src/effect_on_condition.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    }
+}
+
+bool effect_on_condition::activate( dialogue &d ) const
+{
+    if( !has_condition || condition( d ) ) {
+        true_effect.apply( d );
+        return true;
+    } else {
+        false_effect.apply( d );
```
Comment (actual-nh) [inline src/effect_on_condition.cpp:142]: Would it be faster to check has_false_effect before bothering to try?

File: src/effect_on_condition.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        bool activate( dialogue &d ) const;
+        bool check_deactivate() const;
+        void load( const JsonObject &jo, const std::string &src );
+        void finalize();
+        void check() const;
+        effect_on_condition() = default;
+};
+namespace effect_on_conditions
+{
+/** Get all currently loaded generic_preconditions */
```
Comment (actual-nh) [inline src/effect_on_condition.h:55]: The comments are still saying generic_preconditions; should this be changed?

File: src/effect_on_condition.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+const std::vector<effect_on_condition> &get_all();
+/** Finalize all loaded generic_preconditions */
+void finalize_all();
+/** Clear all loaded effects on condition (invalidating any pointers) */
+void reset();
+/** Load effect on condition from JSON definition */
+void load( const JsonObject &jo, const std::string &src );
+/** Checks all loaded from JSON are valid */
+void check_consistency();
+/** sets up the intial queue for a new character */
```
Comment (actual-nh) [inline src/effect_on_condition.h:65]:
```suggestion
/** Sets up the initial queue for a new character */
```

File: src/effect_on_condition.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+void load_new_character();
+/** queue an eoc to happen in the future */
+void queue_effect_on_condition( time_duration duration, effect_on_condition_id eoc );
+/** called every turn to process the queued eocs */
+void process_effect_on_conditions();
+/** called after certain events to test reactivating eocs */
```
Comment (actual-nh) [inline src/effect_on_condition.h:71]:
```suggestion
/** called after certain events to test whether to reactivate eocs */
```

File: src/game.cpp [hunk 6]
```diff
@@ -1611,7 +1613,7 @@ bool game::do_turn()
         ui_manager::redraw();
         refresh_display();
     }
-
+    effect_on_conditions::process_effect_on_conditions();
     if( levz >= 0 && !u.is_underwater() ) {
```
Comment (actual-nh) [inline src/game.cpp:1617]:
```suggestion

    if( levz >= 0 && !u.is_underwater() ) {
```

File: src/init.cpp [hunk 7]
```diff
@@ -700,6 +704,7 @@ void DynamicDataLoader::check_consistency( loading_ui &ui )
             },
             { _( "Vitamins" ), &vitamin::check_consistency },
             { _( "Weather types" ), &weather_types::check_consistency },
+            { _( "Effect on condition" ), &effect_on_conditions::check_consistency },
```
Comment (actual-nh) [inline src/init.cpp:707]:
```suggestion
            { _( "Effect on conditions" ), &effect_on_conditions::check_consistency },
```

**Review by actual-nh — COMMENTED** [review]
Nice!

**Pull request merged**